src/ai/backend/manager/idle.py

```diff
@@ -326,6 +384,164 @@ async def update_app_streaming_status(
+                )
+            )
+            result = await dbconn.execute(query)
+            rows = result.fetchall()
+            kernel_ids = [k["id"] for k in rows]
+        else:
+            kernel_ids = [session_id]
+        current_utilizations = await self.get_current_utilization(kernel_ids, occupied_slots)
+
+        if current_utilizations is None:
+            return True
+
+        # Update utilization time-series data.
+        not_enough_data = False
+        util_series_key = f"session.{session_id}.util_series"
+        raw_util_series = await self._redis.get(util_series_key, encoding=None)
+
+        try:
+            util_series = json.loads(raw_util_series)
+        except (TypeError, json.decoder.JSONDecodeError):
+            util_series = {k: [] for k in self.resource_thresholds}
+        for k in util_series:
+            util_series[k].append(current_utilizations[k])
+            if len(util_series[k]) > window_size:
+                util_series[k].pop(0)
+            else:
+                not_enough_data = True
+        await self._redis.set(
+            util_series_key,
+            json.dumps(util_series),
+            expire=max(86400, int(self.time_window.total_seconds() * 2)),
+        )
+        if not_enough_data:
+            return True
+
+        # Check over-utilized (not to be collected) resources.
+
+        avg_utils = {k: sum(util_series[k]) / len(util_series[k]) for k in util_series}
+        over_utilized = {
+            k: (float(avg_utils[k]) >= float(self.resource_thresholds[k]))
+            for k in self.resource_thresholds
+            if self.resource_thresholds[k] is not None
+        }
+
+        if len(over_utilized) < 1:
+            check_result = True
+        elif self.thresholds_check_operator.lower() == "or":
+            check_result = any(over_utilized.values())
+        else:  # "and" operation is the default
+            check_result = all(over_utilized.values())
+        return check_result
+
+    async def get_current_utilization(
+        self,
+        kernel_ids: Sequence[KernelId],
+        occupied_slots: Mapping[str, Any]
+    ) -> Mapping[str, float] | None:
+        """
+        Return the current utilization key-value pairs of multiple kernels, possibly the
+        components of a cluster session. If there are multiple kernel_ids, this method
+        will return the averaged values over the kernels for each utilization.
+        """
+        try:
+            utilizations = {k: 0.0 for k in self.resource_thresholds}
+            for kernel_id in kernel_ids:
+                raw_live_stat = await self._redis_stat.get(str(kernel_id), encoding=None)
+                live_stat = msgpack.unpackb(raw_live_stat)
+                kernel_utils = {
+                    k: float(nmget(live_stat, f"{k}.pct", 0.0))
+                    for k in self.resource_thresholds
+                }
+                kernel_utils["cpu"] = float(nmget(live_stat, "cpu_util.pct", 0.0))
```

Comment:
I there any reason to change previous `resource-thresholds.cpu_util` Etcd key to `resource-thresholds.cpu`, and adding this line to explicitly convert `cpu_util` in Redis to `cpu`?

Reply:
At that time I was trying to check the execution of idle checker and found bugs while trying to execute idle checker.

According to the history of your latest commit, I found that in config.py there is a _cpu_ key value setting. So, trafaret was giving an error that _cpu is not valid key_. So, I changed in resource-thresholds.cpu_util to resource-thresholds.cpu to match with etcd config.py settings.

For adding explicitly, the Redis live_stat output doesn't contain the cpu key. It contains cpu_util, so that _k: float(nmget(live_stat, f"{k}.pct", 0.0))_ line of code wasn't able to access cpu values under the key _cpu_util_ and cpu values in stream were zero. 

Maybe settings all keys to cpu_util including the config.py to match Redis data stream is better? Then there will be not needed for explicit cpu assignment line.

Reply:
Yes. I think it would be better to change the Etcd key to `cpu_util` than adding a code to handle an exception.

Reply:
Completed making cpu_util consistent